# Lab exercise 2

### 1 - Ambiguity

1a)

In [1]:
import nltk
from nltk.corpus import brown
from collections import defaultdict

In [2]:
brown_words = brown.words()

brown_pos = nltk.pos_tag(brown_words)

In [3]:
tags = defaultdict(set)
freqtag = defaultdict(int)

for word, tag in brown_pos:
    tags[word].add(tag)
    freqtag[tag] += 1

In [4]:
tag_list = [(k, v) for k, v in freqtag.items()]

print(sorted(tag_list, key=lambda x: x[1], reverse=True)[:5])

[('NN', 160733), ('IN', 135982), ('DT', 116201), ('JJ', 81065), ('NNP', 70646)]


``` [('NN', 160733), ('IN', 135982), ('DT', 116201), ('JJ', 81065), ('NNP', 70646)] ```

1b)

In [5]:
amb_words_list = {k: set(v) for k, v in tags.items() if len(v) > 2} 

amb_words = len(amb_words_list.keys())

print(amb_words)

3118


```3118```

1c)

In [6]:
word_count = len(set(brown_words))

In [7]:
prosent = (float(amb_words) / word_count) * 100

print(f"{prosent:.5}%")

5.5622%


```5.5622%```

1d)

In [8]:
dist_words = {k:set(v) for k, v in amb_words_list.items() if len(k) > 4}

In [9]:
for word, tag_list in sorted(dist_words.items(), key=lambda x: len(x[1]), reverse=True)[:5]:
    print("{}: {}".format(word, tag_list))

you're: {'NNP', 'PDT', 'VBZ', 'VBD', 'RB', 'JJ', 'VB', 'VBN', 'JJR', 'CC', 'VBP', 'PRP', 'NNS', 'NN'}
you'll: {'VBN', 'NNP', 'VBZ', 'RB', 'JJ', 'CC', 'VB', 'JJR', 'VBP', 'PRP', 'NNS', 'NN'}
You're: {'UH', 'IN', 'PRP$', 'VBZ', 'FW', 'RB', 'VB', 'CC', 'NNP', 'JJ', 'MD', 'NN'}
that's: {'IN', 'PDT', 'VBZ', 'VBD', 'FW', 'RB', 'JJ', 'VB', 'VBN', 'VBP', 'NN'}
we're: {'IN', 'WP', 'VBZ', 'RB', 'VB', 'JJ', 'VBN', 'VBP', 'NNS', 'NN'}


```you're: {'VBD', 'VBN', 'RB', 'JJ', 'NNP', 'NNS', 'CC', 'VBP', 'JJR', 'PDT', 'PRP', 'VBZ', 'VB', 'NN'}
you'll: {'VBN', 'RB', 'JJ', 'CC', 'NNS', 'NNP', 'VBP', 'JJR', 'PRP', 'VBZ', 'VB', 'NN'}
You're: {'UH', 'IN', 'RB', 'CC', 'NNP', 'PRP$', 'JJ', 'MD', 'VBZ', 'FW', 'VB', 'NN'}
that's: {'VBN', 'IN', 'JJ', 'RB', 'VBP', 'NN', 'PDT', 'VBZ', 'FW', 'VB', 'VBD'}
we're: {'VBN', 'IN', 'RB', 'JJ', 'NNS', 'VBP', 'WP', 'VBZ', 'VB', 'NN'}```

Choose the word "that's".

In [10]:
from nltk.text import Text
from nltk import sent_tokenize, word_tokenize
chosen_word = "that's"

In [11]:
tags_to_find = set(dist_words[chosen_word])
print(tags_to_find)
sentences = []
for sent in brown.sents():
    if chosen_word in sent:
        m = {w:t for w, t in nltk.pos_tag(sent)}
        if m[chosen_word] in tags_to_find:
            sentences.append((m[chosen_word], sent))
            tags_to_find.remove(m[chosen_word])


{'IN', 'PDT', 'VBZ', 'VBD', 'FW', 'RB', 'JJ', 'VB', 'VBN', 'VBP', 'NN'}


In [12]:
for tag, s in sentences:
    print(tag, "-", " ".join(s))

NN - And that's meant as a boost , not a knock .
PDT - Comic Gary Morton signed to play the Living Room here Dec. 18 , because that's the only time his heart , Lucille Ball , can come along .
VB - He called the Cuban tractor plan an outright blackmail action , and noted that in war `` you can't buy yourself out and that's what we're trying to do '' .
VBN - If that's all the Romans did , it's a surprise to me that Rome fell .
VBD - If most of them weren't exactly specific -- well , that's the way it is in life , I guess .
VBZ - Although he is not graced with the subtleties of romantic technique , that's not what an ex-prize fighter is supposed to have , anyway .
JJ - He should have said `` jittery '' , for that's what we are .
VBP - Well , whatever you have , that's it ! !
IN - Uh huh , we think , looking at them , so that's the Parthenon .
RB - `` I'll bet that's as close as you've been to a man since you were a baby '' , Wilson said .
FW - He did it because he knows for each guy he pu

d)
```
NN - And that's meant as a boost , not a knock .
PDT - Comic Gary Morton signed to play the Living Room here Dec. 18 , because that's the only time his heart , Lucille Ball , can come along .
VB - He called the Cuban tractor plan an outright blackmail action , and noted that in war `` you can't buy yourself out and that's what we're trying to do '' .
VBN - If that's all the Romans did , it's a surprise to me that Rome fell .
VBD - If most of them weren't exactly specific -- well , that's the way it is in life , I guess .
VBZ - Although he is not graced with the subtleties of romantic technique , that's not what an ex-prize fighter is supposed to have , anyway .
JJ - He should have said `` jittery '' , for that's what we are .
VBP - Well , whatever you have , that's it ! !
IN - Uh huh , we think , looking at them , so that's the Parthenon .
RB - `` I'll bet that's as close as you've been to a man since you were a baby '' , Wilson said .
FW - He did it because he knows for each guy he puts out of commission that's one less who might take his job away later on .
```

e)
To reduce ambiguity we could look at the words and tags coming before as well as the current word and tag to improve the prediction

### 2 - Training a tagger

In [13]:
from nltk.corpus import nps_chat
from nltk.tag import DefaultTagger, RegexpTagger, UnigramTagger, TrigramTagger, BigramTagger


nps_text = Text(nps_chat.words())
brown_text = Text(brown.words())

nps_tagged = nps_chat.tagged_posts()
brown_tagged = brown.tagged_sents()

In [14]:
brown_most_freq_tag = "NN"
freqtag = defaultdict(int)
for words in nps_tagged:
    for word, tag in words:
        freqtag[tag] += 1
tag_list = [(k, v) for k, v in freqtag.items()]

nps_most_freq_tag = sorted(tag_list, key=lambda x: x[1], reverse=True)[0][0]

brown_word_count = len(brown_text)
nps_word_count = len(nps_text)

training_sets = [
    (brown_tagged[0:int(brown_word_count * 0.9)], nps_tagged[0:int(nps_word_count * 0.1)]),
    (brown_tagged[0:int(brown_word_count * 0.5)], nps_tagged[0:int(nps_word_count * 0.5)]),
    (nps_tagged[0:int(nps_word_count * 0.9)], brown_tagged[0:int(brown_word_count * 0.1)]),
    (nps_tagged[0:int(nps_word_count * 0.5)], brown_tagged[0:int(brown_word_count * 0.5)])
]

a)

In [15]:
brown_default_tagger = DefaultTagger(brown_most_freq_tag)
nps_default_tagger = DefaultTagger(nps_most_freq_tag)

print(brown_default_tagger.evaluate(brown_tagged))
print(brown_default_tagger.evaluate(nps_tagged))
print(nps_default_tagger.evaluate(brown_tagged))
print(nps_default_tagger.evaluate(nps_tagged))

C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:4: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  after removing the cwd from sys.path.


0.13130472824476916


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:5: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  """


0.08871361919573428


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:6: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  


0.000523599887012656


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:7: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  import sys


0.11373028215952011


* 0.13130472824476916
* 0.08871361919573428
* 0.000523599887012656
* 0.11373028215952011

In [16]:
def train_tagger(training_set, test_set, default_tagger):
    backoff = RegexpTagger([
        (r'^-?[0-9]+(.[0-9]+)?$', 'CD'),   # cardinal numbers
        (r'(The|the|A|a|An|an)$', 'AT'),   # articles
        (r'.*able$', 'JJ'),                # adjectives
        (r'.*ness$', 'NN'),                # nouns formed from adjectives
        (r'.*ly$', 'RB'),                  # adverbs
        (r'.*s$', 'NNS'),                  # plural nouns
        (r'.*ing$', 'VBG'),                # gerunds
        (r'.*ed$', 'VBD'),                 # past tense verbs
        ],
        backoff=default_tagger
    )
    print("     RegExp accuracy:", backoff.accuracy(test_set))

    backoff = UnigramTagger(
        training_set,
        backoff=backoff
    )
    print("    Unigram accuracy:", backoff.accuracy(test_set))
    
    backoff = BigramTagger(
        training_set,
        backoff=backoff
    )
    print("    Bigram accuracy:", backoff.accuracy(test_set))
       
    backoff = TrigramTagger(
        training_set,
        backoff=backoff
    )
    print("    Trigram accuracy:", backoff.accuracy(test_set))
    
    return backoff

taggers = list()
for i, (training, test) in enumerate(training_sets):
    print(f"Training on set {i}")
    taggers.append(train_tagger(training, test, brown_default_tagger))

Training on set 0


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:14: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  


RegExp accuracy: 0.13261984612349575


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:20: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.


Unigram accuracy: 0.3662458078516473


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:26: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.


Bigram accuracy: 0.36402643519431843


C:\Users\andri\anaconda3\lib\site-packages\ipykernel_launcher.py:32: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.


Trigram accuracy: 0.3619550207141448
Training on set 1
RegExp accuracy: 0.1325483225949789
Unigram accuracy: 0.3631415241057543
Bigram accuracy: 0.36278604754499
Trigram accuracy: 0.36145301044212397
Training on set 2
RegExp accuracy: 0.2961835768761755
Unigram accuracy: 0.5366382131464909
Bigram accuracy: 0.5368147558715527
Trigram accuracy: 0.5369637407078244
Training on set 3
RegExp accuracy: 0.2961835768761755
Unigram accuracy: 0.5366382131464909
Bigram accuracy: 0.5368147558715527
Trigram accuracy: 0.5369637407078244


```
Training on set 0
    RegExp accuracy: 0.13261984612349575
    Unigram accuracy: 0.3662458078516473
    Bigram accuracy: 0.36402643519431843
    Trigram accuracy: 0.3619550207141448
Training on set 1
    RegExp accuracy: 0.1325483225949789
    Unigram accuracy: 0.3631415241057543
    Bigram accuracy: 0.36278604754499
    Trigram accuracy: 0.36145301044212397
Training on set 2
    RegExp accuracy: 0.2961835768761755
    Unigram accuracy: 0.5366382131464909
    Bigram accuracy: 0.5368147558715527
    Trigram accuracy: 0.5369637407078244
Training on set 3
    RegExp accuracy: 0.2961835768761755
    Unigram accuracy: 0.5366382131464909
    Bigram accuracy: 0.5368147558715527
    Trigram accuracy: 0.5369637407078244
```

The taggers are chosen in the order of increased complexity.
It seems that running the Ngram taggers in order does not impacy accuracy that much. So each more complex tagger falls back to a simpler one.

c)
Chose a 50/50 split

In [18]:
#This is the tagger with 50/50 split
tagger = taggers[1]
golden_standard = training_sets[1][1]
print(tagger.evaluate_per_tag(golden_standard, truncate=5, sort_by_count=True))

Tag | Prec.  | Recall | F-measure
----+--------+--------+-----------
 UH | 0.9928 | 0.1338 | 0.2358
NNP | 0.0000 | 0.0000 | 0.0000
 VB | 0.6065 | 0.2895 | 0.3920
 NN | 0.1853 | 0.8520 | 0.3045
PRP | 0.0000 | 0.0000 | 0.0000



```
Tag | Prec.  | Recall | F-measure
----+--------+--------+-----------
 UH | 0.9928 | 0.1338 | 0.2358
NNP | 0.0000 | 0.0000 | 0.0000
 VB | 0.6065 | 0.2895 | 0.3920
 NN | 0.1853 | 0.8520 | 0.3045
PRP | 0.0000 | 0.0000 | 0.0000
```

d)

In [22]:
fd = nltk.FreqDist(brown_words)
cfd = nltk.ConditionalFreqDist(brown.tagged_words())
most_freq_words = fd.most_common(200)
likely_tags = {word: cfd[word].max() for (word, _) in most_freq_words}

baseline_tagger = UnigramTagger(model=likely_tags, backoff=brown_default_tagger)

print("Baseline lookup tagger accuracy:", baseline_tagger.accuracy(golden_standard))


Baseline lookup tagger accuracy: 0.253143745834259


e)

In [30]:
from nltk.corpus import gutenberg
import random
random.seed(101)

gutenberg_sents =random.choices(gutenberg.sents(), k=10)

In [32]:
for s in baseline_tagger.tag_sents(gutenberg_sents):
    print(s)
    print()


[('Separate', 'NN'), ('tall', 'NN'), ('flames', 'NN'), ('shot', 'NN'), ('up', 'RP'), ('and', 'CC'), ('spread', 'NN'), ('out', 'RP'), ('above', 'NN'), ('them', 'PPO'), ('like', 'CS'), ('the', 'AT'), ('fiery', 'NN'), ('cloisters', 'NN'), ('of', 'IN'), ('some', 'DTI'), ('infernal', 'NN'), ('cathedral', 'NN'), (',', ','), ('or', 'CC'), ('like', 'CS'), ('a', 'AT'), ('grove', 'NN'), ('of', 'IN'), ('red', 'NN'), ('tropical', 'NN'), ('trees', 'NN'), ('in', 'IN'), ('the', 'AT'), ('garden', 'NN'), ('of', 'IN'), ('the', 'AT'), ('devil', 'NN'), ('.', '.')]

[('8', 'NN'), (':', ':'), ('7', 'NN'), ('And', 'CC'), ('he', 'PPS'), ('put', 'NN'), ('upon', 'IN'), ('him', 'PPO'), ('the', 'AT'), ('coat', 'NN'), (',', ','), ('and', 'CC'), ('girded', 'NN'), ('him', 'PPO'), ('with', 'IN'), ('the', 'AT'), ('girdle', 'NN'), (',', ','), ('and', 'CC'), ('clothed', 'NN'), ('him', 'PPO'), ('with', 'IN'), ('the', 'AT'), ('robe', 'NN'), (',', ','), ('and', 'CC'), ('put', 'NN'), ('the', 'AT'), ('ephod', 'NN'), ('upon',

```
[('Separate', 'NN'), ('tall', 'NN'), ('flames', 'NN'), ('shot', 'NN'), ('up', 'RP'), ('and', 'CC'), ('spread', 'NN'), ('out', 'RP'), ('above', 'NN'), ('them', 'PPO'), ('like', 'CS'), ('the', 'AT'), ('fiery', 'NN'), ('cloisters', 'NN'), ('of', 'IN'), ('some', 'DTI'), ('infernal', 'NN'), ('cathedral', 'NN'), (',', ','), ('or', 'CC'), ('like', 'CS'), ('a', 'AT'), ('grove', 'NN'), ('of', 'IN'), ('red', 'NN'), ('tropical', 'NN'), ('trees', 'NN'), ('in', 'IN'), ('the', 'AT'), ('garden', 'NN'), ('of', 'IN'), ('the', 'AT'), ('devil', 'NN'), ('.', '.')]

[('8', 'NN'), (':', ':'), ('7', 'NN'), ('And', 'CC'), ('he', 'PPS'), ('put', 'NN'), ('upon', 'IN'), ('him', 'PPO'), ('the', 'AT'), ('coat', 'NN'), (',', ','), ('and', 'CC'), ('girded', 'NN'), ('him', 'PPO'), ('with', 'IN'), ('the', 'AT'), ('girdle', 'NN'), (',', ','), ('and', 'CC'), ('clothed', 'NN'), ('him', 'PPO'), ('with', 'IN'), ('the', 'AT'), ('robe', 'NN'), (',', ','), ('and', 'CC'), ('put', 'NN'), ('the', 'AT'), ('ephod', 'NN'), ('upon', 'IN'), ('him', 'PPO'), (',', ','), ('and', 'CC'), ('he', 'PPS'), ('girded', 'NN'), ('him', 'PPO'), ('with', 'IN'), ('the', 'AT'), ('curious', 'NN'), ('girdle', 'NN'), ('of', 'IN'), ('the', 'AT'), ('ephod', 'NN'), (',', ','), ('and', 'CC'), ('bound', 'NN'), ('it', 'PPS'), ('unto', 'NN'), ('him', 'PPO'), ('therewith', 'NN'), ('.', '.')]

[('O', 'NN'), ('to', 'TO'), ('be', 'BE'), ('yielded', 'NN'), ('to', 'TO'), ('you', 'PPSS'), ('whoever', 'NN'), ('you', 'PPSS'), ('are', 'BER'), (',', ','), ('and', 'CC'), ('you', 'PPSS'), ('to', 'TO'), ('be', 'BE'), ('yielded', 'NN'), ('to', 'TO'), ('me', 'PPO'), ('in', 'IN'), ('defiance', 'NN'), ('of', 'IN'), ('the', 'AT'), ('world', 'NN'), ('!', '.')]

[('Me', 'NN'), ('thinkes', 'NN'), ('it', 'PPS'), ('is', 'BEZ'), ('like', 'CS'), ('a', 'AT'), ('Weazell', 'NN')]

[('2', 'NN'), (':', ':'), ('9', 'NN'), ('He', 'PPS'), ('that', 'CS'), ('saith', 'NN'), ('he', 'PPS'), ('is', 'BEZ'), ('in', 'IN'), ('the', 'AT'), ('light', 'NN'), (',', ','), ('and', 'CC'), ('hateth', 'NN'), ('his', 'PP$'), ('brother', 'NN'), (',', ','), ('is', 'BEZ'), ('in', 'IN'), ('darkness', 'NN'), ('even', 'RB'), ('until', 'NN'), ('now', 'RB'), ('.', '.')]

[('He', 'PPS'), ('chose', 'NN'), ('his', 'PP$'), ('place', 'NN'), ('well', 'RB'), (',', ','), ('and', 'CC'), ('waited', 'NN'), ('nearly', 'NN'), ('all', 'ABN'), ('the', 'AT'), ('evening', 'NN'), (',', ','), ('offering', 'NN'), ('his', 'PP$'), ('fossils', 'NN'), ('with', 'IN'), ('great', 'JJ'), ('assiduity', 'NN'), ('to', 'TO'), ('every', 'NN'), ('passenger', 'NN'), (';', '.'), ('but', 'CC'), ('not', '*'), ('one', 'CD'), ('person', 'NN'), ('bought', 'NN'), ('any', 'DTI'), ('.', '.')]

[('22', 'NN'), (':', ':'), ('4', 'NN'), ('Thou', 'NN'), ('shalt', 'NN'), ('not', '*'), ('see', 'VB'), ('thy', 'NN'), ('brother', 'NN'), ("'", 'NN'), ('s', 'NN'), ('ass', 'NN'), ('or', 'CC'), ('his', 'PP$'), ('ox', 'NN'), ('fall', 'NN'), ('down', 'RP'), ('by', 'IN'), ('the', 'AT'), ('way', 'NN'), (',', ','), ('and', 'CC'), ('hide', 'NN'), ('thyself', 'NN'), ('from', 'IN'), ('them', 'PPO'), (':', ':'), ('thou', 'NN'), ('shalt', 'NN'), ('surely', 'NN'), ('help', 'NN'), ('him', 'PPO'), ('to', 'TO'), ('lift', 'NN'), ('them', 'PPO'), ('up', 'RP'), ('again', 'RB'), ('.', '.')]

[('21', 'NN'), (':', ':'), ('20', 'NN'), ('And', 'CC'), ('the', 'AT'), ('families', 'NN'), ('of', 'IN'), ('the', 'AT'), ('children', 'NN'), ('of', 'IN'), ('Kohath', 'NN'), (',', ','), ('the', 'AT'), ('Levites', 'NN'), ('which', 'WDT'), ('remained', 'NN'), ('of', 'IN'), ('the', 'AT'), ('children', 'NN'), ('of', 'IN'), ('Kohath', 'NN'), (',', ','), ('even', 'RB'), ('they', 'PPSS'), ('had', 'HVD'), ('the', 'AT'), ('cities', 'NN'), ('of', 'IN'), ('their', 'PP$'), ('lot', 'NN'), ('out', 'RP'), ('of', 'IN'), ('the', 'AT'), ('tribe', 'NN'), ('of', 'IN'), ('Ephraim', 'NN'), ('.', '.')]

[('11', 'NN'), (':', ':'), ('13', 'NN'), ('And', 'CC'), ('his', 'PP$'), ('brethren', 'NN'), (',', ','), ('chief', 'NN'), ('of', 'IN'), ('the', 'AT'), ('fathers', 'NN'), (',', ','), ('two', 'CD'), ('hundred', 'NN'), ('forty', 'NN'), ('and', 'CC'), ('two', 'CD'), (':', ':'), ('and', 'CC'), ('Amashai', 'NN'), ('the', 'AT'), ('son', 'NN'), ('of', 'IN'), ('Azareel', 'NN'), (',', ','), ('the', 'AT'), ('son', 'NN'), ('of', 'IN'), ('Ahasai', 'NN'), (',', ','), ('the', 'AT'), ('son', 'NN'), ('of', 'IN'), ('Meshillemoth', 'NN'), (',', ','), ('the', 'AT'), ('son', 'NN'), ('of', 'IN'), ('Immer', 'NN'), (',', ','), ('11', 'NN'), (':', ':'), ('14', 'NN'), ('And', 'CC'), ('their', 'PP$'), ('brethren', 'NN'), (',', ','), ('mighty', 'NN'), ('men', 'NNS'), ('of', 'IN'), ('valour', 'NN'), (',', ','), ('an', 'AT'), ('hundred', 'NN'), ('twenty', 'NN'), ('and', 'CC'), ('eight', 'NN'), (':', ':'), ('and', 'CC'), ('their', 'PP$'), ('overseer', 'NN'), ('was', 'BEDZ'), ('Zabdiel', 'NN'), (',', ','), ('the', 'AT'), ('son', 'NN'), ('of', 'IN'), ('one', 'CD'), ('of', 'IN'), ('the', 'AT'), ('great', 'JJ'), ('men', 'NNS'), ('.', '.')]

[('But', 'CC'), ('Bursal', 'NN'), ("'", 'NN'), ('s', 'NN'), ('my', 'PP$'), ('witness', 'NN'), ('--', '--')]

```

f)

### 3 - Tagging with probabilities

a)

In [73]:
import nltk
from nltk.corpus import brown
from nltk.probability import MLEProbDist
from nltk.util import ngrams
# see https://nltk.readthedocs.io/en/latest/api/nltk.html
# define distinguishable start/end tuples of tag/word
# used to mark sentences
START = ("START", "START")
END = ("END", "END")

def get_tags(corpus):
    tags_words = []
    for sent in corpus.tagged_sents():
        tags_words.append(START)
        # shorten tags to 2 characters each for simplicity
        tags_words.extend([(tag[:2], word) for (word, tag) in sent])
        tags_words.append(END)

    return tags_words

def probDist(corpus, probability_distribution, tag_observation_fn):
    tag_words = get_tags(corpus)
    tags = [tag for (tag, _) in tag_words]
    # conditional frequency distribution over tag/word
    cfd_tagwords = nltk.ConditionalFreqDist(tag_words)
    cpd_tagwords = nltk.ConditionalProbDist(cfd_tagwords, probability_distribution)
    # conditional frequency distribution of observations:
    cfd_tags = nltk.ConditionalFreqDist(tag_observation_fn(tags))
    cpd_tags = nltk.ConditionalProbDist(cfd_tags, probability_distribution)
    
    return cpd_tagwords, cpd_tags

def task3a():
    corpus = brown
    # maximum likelihood estimate to create a probability distribution 
    probability_distribution = nltk.MLEProbDist
    # a function to create tag observations. Hint: can we observe anything with unigrams?
    tag_observation_fn = lambda tags: ngrams(tags, 2)
    return probDist(corpus, probability_distribution, tag_observation_fn)
    
def prettify(prob):
    return "{}%".format(round(prob * 100, 4))
    
def task3b():
    tagwords, tags = task3a()
    prob_verb_is_run = tagwords["VB"].prob("run")  # IMPLEMENT
    prob_v_follows_p = tags["PP"].prob("VB")  # IMPLEMENT
    print("Prob. of a Verb(VB) being 'run' is", prettify(prob_verb_is_run))
    print("Prob. of a Preposition(PP) being followed by a Verb(VB) is", prettify(prob_v_follows_p))

In [74]:
task3b()

Prob. of a Verb(VB) being 'run' is 0.1329%
Prob. of a Preposition(PP) being followed by a Verb(VB) is 25.1591%


b)

Prob. of a Verb(VB) being 'run' is 0.1329%

Prob. of a Preposition(PP) being followed by a Verb(VB) is 25.1591%

In [77]:
tagwords, tags = task3a()

In [93]:
for tag in ["NN", "VB", "JJ"]:
    print(tag + ":\n  ", ", ".join(map(lambda e: e[0], tagwords[tag].freqdist().most_common(10))))

NN:
   time, man, Af, years, way, people, men, world, life, year
VB:
   said, made, make, see, get, know, came, used, go, come
JJ:
   new, such, own, good, great, New, old, American, small, long


c)
```
NN:
   time, man, Af, years, way, people, men, world, life, year
VB:
   said, made, make, see, get, know, came, used, go, come
JJ:
   new, such, own, good, great, New, old, American, small, long
```

In [94]:
sequence = "I can code some code"

d)

I am not sure how to solve this one.